lib/bindings/AMQPBinding.js

```diff
@@ -135,17 +135,24 @@ function start(callback) {
                     }
                 });
                 conn.on('close', function() {
-                    config.getLogger().error(context, 'reconnecting');
-                    if (numRetried <= retries) {
-                        numRetried++;
-                        return setTimeout(createConnection,
-                                          retryTime * 1000,
-                                          callback);
+                    // If amqpConn is null, the connection has been closed on purpose
```

Comment:
As far as I understand, this change is not due to changes in the amqp library API, as it has not been updated in this PR (https://github.com/telefonicaid/iotagent-json/pull/339/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2L51).

Thus, could you elaborate on this change, pls?

Reply:
After updating `async` module, tests related to AMQP binding started to fail due to a `Callback already called error`.

I checked the code of https://github.com/telefonicaid/iotagent-json/blob/master/lib/bindings/AMQPBinding.js and identify some reasons:
- In https://github.com/telefonicaid/iotagent-json/blob/master/lib/bindings/AMQPBinding.js#L176 and https://github.com/telefonicaid/iotagent-json/blob/master/lib/bindings/AMQPBinding.js#L177, the same `callback` is passed to the two `assertQueue` functions. This will cause that both of them call to the same `callback` function...
- In https://github.com/telefonicaid/iotagent-json/blob/master/lib/bindings/AMQPBinding.js#137, if the connection closes, this part of the code tries to reconnect using return `setTimeout(createConnection,retryTime * 1000,callback)`. As you can see, we are passing again the callback to setTimeout, but this callback was already called when the connection was initially established (https://github.com/telefonicaid/iotagent-json/blob/master/lib/bindings/AMQPBinding.js#L148). Then, if the reconnection succeeds, we will have a `Callback already called` error. For this reason, I use directly return `setTimeout(createConnection, retryTime * 1000);` without callback.
- Finally, https://github.com/telefonicaid/iotagent-json/blob/master/lib/bindings/AMQPBinding.js will be executed also when the connection is closed by `close` function of https://github.com/telefonicaid/iotagent-json/blob/master/lib/bindings/AMQPBinding.js#L209, and the code will try to reconnect... I think that this does not make sense. For that reason, in `close` function I set amqpConn and check it in `close` event handler.

Reply:
Thank you for so detailed explanation :)

Everything seems to be ok. My only doubt if this is could be considered an existing issue in the AMQP binding logic (that the update of aync has raised but that has been always there) so the fix should be mentioned in the CNR. Something like:

```
Fix: AMQP callback over-calling
```

or maybe something more precise.

Up to you :) Either close this comment thread with NTC (if CNR is ok as it is now) or with Fixed in (if CNR needs to be modified)

Reply:
Fixed with 9d73a9bc73942890c0e31901649b01d61b581deb